File: TestSpawning.cpp
// Simple standalone test to verify spawning logic
#include <iostream>
#include <vector>

class GridVisualizerTest {
public:
    static void TestSpawning() {
        std::cout << "========== GRID VISUALIZER SPAWN TEST ==========\n";
        
        // Test configuration
        const int GridSizeX = 100;
        const int GridSizeY = 100;
        const int RoomTiles = 3;
        const int RoomSpacing = 5;
        const float CellSize = 1000.0f;
        
        // Track what should be spawned
        int ExpectedFloors = 0;
        int ExpectedWalls = 0;
        
        // Simulate grid initialization
        std::vector<int> TestGrid(GridSizeX * GridSizeY, 0);
        
        // Simulate placing boundary walls
        for (int X = 0; X < GridSizeX; ++X) {
            TestGrid[X] = 2; // Top edge wall
            TestGrid[(GridSizeY - 1) * GridSizeX + X] = 2; // Bottom edge wall
            ExpectedWalls += 2;
        }
        
        for (int Y = 0; Y < GridSizeY; ++Y) {
            TestGrid[Y * GridSizeX] = 2; // Left edge wall
            TestGrid[Y * GridSizeX + (GridSizeX - 1)] = 2; // Right edge wall
            ExpectedWalls += 2;
        }
        
        // Adjust for corners counted twice
        ExpectedWalls -= 4;
        
        std::cout << "Boundary walls placed: " << ExpectedWalls << "\n";
        
        // Simulate placing 25 rooms in a 5x5 grid
        int RoomsPlaced = 0;
        for (int RoomY = 0; RoomY < 5; ++RoomY) {
            for (int RoomX = 0; RoomX < 5; ++RoomX) {
                int StartX = RoomX * RoomSpacing + 1;
                int StartY = RoomY * RoomSpacing + 1;
                
                // Place floor tiles (3x3 interior)
                for (int X = StartX + 1; X < StartX + 1 + RoomTiles; ++X) {
                    for (int Y = StartY + 1; Y < StartY + 1 + RoomTiles; ++Y) {
                        if (X < GridSizeX && Y < GridSizeY) {
                            TestGrid[Y * GridSizeX + X] = 1; // Floor
                            ExpectedFloors++;
                        }
                    }
                }
                
                // Place room walls
                // Top and bottom walls
                for (int X = StartX; X <= StartX + RoomTiles + 1; ++X) {
                    if (X < GridSizeX) {
                        if (StartY < GridSizeY) {
                            TestGrid[StartY * GridSizeX + X] = 2; // Top wall
                            ExpectedWalls++;
                        }
                        if (StartY + RoomTiles + 1 < GridSizeY) {
                            TestGrid[(StartY + RoomTiles + 1) * GridSizeX + X] = 2; // Bottom wall
                            ExpectedWalls++;
                        }
                    }
                }
                
                // Left and right walls
                for (int Y = StartY; Y <= StartY + RoomTiles + 1; ++Y) {
                    if (Y < GridSizeY) {
                        if (StartX < GridSizeX) {
                            TestGrid[Y * GridSizeX + StartX] = 2; // Left wall
                            ExpectedWalls++;
                        }
                        if (StartX + RoomTiles + 1 < GridSizeX) {
                            TestGrid[Y * GridSizeX + (StartX + RoomTiles + 1)] = 2; // Right wall
                            ExpectedWalls++;
                        }
                    }
                }
                
                RoomsPlaced++;
            }
        }
        
        // Count actual cells in grid
        int ActualFloors = 0;
        int ActualWalls = 0;
        int ActualEmpty = 0;
        
        for (size_t i = 0; i < TestGrid.size(); ++i) {
            if (TestGrid[i] == 1) ActualFloors++;
            else if (TestGrid[i] == 2) ActualWalls++;
            else ActualEmpty++;
        }
        
        // Log results
        std::cout << "\n========== TEST RESULTS ==========\n";
        std::cout << "Grid Size: " << GridSizeX << "x" << GridSizeY << " = " << TestGrid.size() << " cells\n";
        std::cout << "Rooms Placed: " << RoomsPlaced << " (5x5 grid)\n";
        std::cout << "Room Size: " << RoomTiles << "x" << RoomTiles << " tiles interior\n";
        std::cout << "Cell Size: " << CellSize << " units\n\n";
        
        std::cout << "Floor Tiles: " << ActualFloors << "\n";
        std::cout << "Wall Tiles: " << ActualWalls << "\n";
        std::cout << "Empty Cells: " << ActualEmpty << "\n";
        std::cout << "Total: " << (ActualFloors + ActualWalls + ActualEmpty) << "\n\n";
        
        // Calculate world space dimensions
        float RoomWorldSize = RoomTiles * CellSize;
        float TotalWorldWidth = GridSizeX * CellSize;
        float TotalWorldHeight = GridSizeY * CellSize;
        
        std::cout << "========== WORLD DIMENSIONS ==========\n";
        std::cout << "Each Room Interior: " << RoomWorldSize << "x" << RoomWorldSize << " units\n";
        std::cout << "Total Dungeon Size: " << TotalWorldWidth << "x" << TotalWorldHeight << " units\n";
        std::cout << "Mannequin Scale: Each room is " << (RoomWorldSize / 183.0f) 
                  << "x" << (RoomWorldSize / 183.0f) << " mannequins\n";
        
        // Verify spawning expectations
        bool bFloorTestPassed = ActualFloors > 0 && ActualFloors == 25 * 9; // 25 rooms * 9 tiles each
        bool bWallTestPassed = ActualWalls > 0;
        bool bGridTestPassed = (ActualFloors + ActualWalls + ActualEmpty) == static_cast<int>(TestGrid.size());
        
        std::cout << "\n========== VALIDATION ==========\n";
        std::cout << "Floor Generation: " << (bFloorTestPassed ? "PASS" : "FAIL") 
                  << " (Expected " << (25 * 9) << " floors for 25 rooms)\n";
        std::cout << "Wall Generation: " << (bWallTestPassed ? "PASS" : "FAIL") 
                  << " (" << ActualWalls << " walls generated)\n";
        std::cout << "Grid Integrity: " << (bGridTestPassed ? "PASS" : "FAIL") << "\n";
        
        // Test mesh scaling calculations
        float PlaneScale = CellSize / 100.0f;
        float CubeScaleXY = CellSize / 100.0f;
        float CubeScaleZ = 500.0f / 100.0f; // WallHeight / 100
        
        std::cout << "\n========== MESH SCALING ==========\n";
        std::cout << "Plane Scale: " << PlaneScale << " x " << PlaneScale << " x 1.0\n";
        std::cout << "Cube Scale: " << CubeScaleXY << " x " << CubeScaleXY << " x " << CubeScaleZ << "\n";
        std::cout << "(Base mesh is 100x100 units, scaled to " << CellSize << "x" << CellSize << ")\n";
        
        std::cout << "\n========== MESH INSTANCES TO SPAWN ==========\n";
        std::cout << "Floor meshes (planes): " << ActualFloors << " instances\n";
        std::cout << "Wall meshes (cubes): " << ActualWalls << " instances\n";
        std::cout << "Total mesh instances: " << (ActualFloors + ActualWalls) << "\n";
        
        std::cout << "\n========================================\n";
        
        if (bFloorTestPassed && bWallTestPassed && bGridTestPassed) {
            std::cout << "\n*** ALL TESTS PASSED ***\n";
            std::cout << "The spawning logic should generate:\n";
            std::cout << "- " << ActualFloors << " floor plane instances\n";
            std::cout << "- " << ActualWalls << " wall cube instances\n";
        } else {
            std::cout << "\n*** SOME TESTS FAILED ***\n";
        }
    }
};

int main() {
    GridVisualizerTest::TestSpawning();
    return 0;
}

File: testing/python/test_spawning.py
#!/usr/bin/env python3
"""
Backend test to verify GridDungeonVisualizer spawning logic
"""

def test_spawning():
    print("========== GRID VISUALIZER SPAWN TEST ==========")
    
    # Test configuration
    grid_size_x = 100
    grid_size_y = 100
    room_tiles = 3
    room_spacing = 5
    cell_size = 1000.0
    
    # Track what should be spawned
    expected_floors = 0
    expected_walls = 0
    
    # Simulate grid initialization
    test_grid = [0] * (grid_size_x * grid_size_y)
    
    # Simulate placing boundary walls
    for x in range(grid_size_x):
        test_grid[x] = 2  # Top edge wall
        test_grid[(grid_size_y - 1) * grid_size_x + x] = 2  # Bottom edge wall
        expected_walls += 2
    
    for y in range(grid_size_y):
        test_grid[y * grid_size_x] = 2  # Left edge wall
        test_grid[y * grid_size_x + (grid_size_x - 1)] = 2  # Right edge wall
        expected_walls += 2
    
    # Adjust for corners counted twice
    expected_walls -= 4
    
    print(f"Boundary walls placed: {expected_walls}")
    
    # Simulate placing 25 rooms in a 5x5 grid
    rooms_placed = 0
    for room_y in range(5):
        for room_x in range(5):
            start_x = room_x * room_spacing + 1
            start_y = room_y * room_spacing + 1
            
            # Place floor tiles (3x3 interior)
            for x in range(start_x + 1, start_x + 1 + room_tiles):
                for y in range(start_y + 1, start_y + 1 + room_tiles):
                    if x < grid_size_x and y < grid_size_y:
                        test_grid[y * grid_size_x + x] = 1  # Floor
                        expected_floors += 1
            
            # Place room walls
            # Top and bottom walls
            for x in range(start_x, start_x + room_tiles + 2):
                if x < grid_size_x:
                    if start_y < grid_size_y:
                        test_grid[start_y * grid_size_x + x] = 2  # Top wall
                        expected_walls += 1
                    if start_y + room_tiles + 1 < grid_size_y:
                        test_grid[(start_y + room_tiles + 1) * grid_size_x + x] = 2  # Bottom wall
                        expected_walls += 1
            
            # Left and right walls
            for y in range(start_y, start_y + room_tiles + 2):
                if y < grid_size_y:
                    if start_x < grid_size_x:
                        test_grid[y * grid_size_x + start_x] = 2  # Left wall
                        expected_walls += 1
                    if start_x + room_tiles + 1 < grid_size_x:
                        test_grid[y * grid_size_x + (start_x + room_tiles + 1)] = 2  # Right wall
                        expected_walls += 1
            
            rooms_placed += 1
    
    # Count actual cells in grid
    actual_floors = sum(1 for cell in test_grid if cell == 1)
    actual_walls = sum(1 for cell in test_grid if cell == 2)
    actual_empty = sum(1 for cell in test_grid if cell == 0)
    
    # Log results
    print("\n========== TEST RESULTS ==========")
    print(f"Grid Size: {grid_size_x}x{grid_size_y} = {len(test_grid)} cells")
    print(f"Rooms Placed: {rooms_placed} (5x5 grid)")
    print(f"Room Size: {room_tiles}x{room_tiles} tiles interior")
    print(f"Cell Size: {cell_size} units\n")
    
    print(f"Floor Tiles: {actual_floors}")
    print(f"Wall Tiles: {actual_walls}")
    print(f"Empty Cells: {actual_empty}")
    print(f"Total: {actual_floors + actual_walls + actual_empty}\n")
    
    # Calculate world space dimensions
    room_world_size = room_tiles * cell_size
    total_world_width = grid_size_x * cell_size
    total_world_height = grid_size_y * cell_size
    
    print("========== WORLD DIMENSIONS ==========")
    print(f"Each Room Interior: {room_world_size}x{room_world_size} units")
    print(f"Total Dungeon Size: {total_world_width}x{total_world_height} units")
    print(f"Mannequin Scale: Each room is {room_world_size / 183.0:.1f}x{room_world_size / 183.0:.1f} mannequins")
    
    # Verify spawning expectations
    floor_test_passed = actual_floors > 0 and actual_floors == 25 * 9  # 25 rooms * 9 tiles each
    wall_test_passed = actual_walls > 0
    grid_test_passed = (actual_floors + actual_walls + actual_empty) == len(test_grid)
    
    print("\n========== VALIDATION ==========")
    print(f"Floor Generation: {'PASS' if floor_test_passed else 'FAIL'} (Expected {25 * 9} floors for 25 rooms)")
    print(f"Wall Generation: {'PASS' if wall_test_passed else 'FAIL'} ({actual_walls} walls generated)")
    print(f"Grid Integrity: {'PASS' if grid_test_passed else 'FAIL'}")
    
    # Test mesh scaling calculations
    plane_scale = cell_size / 100.0
    cube_scale_xy = cell_size / 100.0
    cube_scale_z = 500.0 / 100.0  # WallHeight / 100
    
    print("\n========== MESH SCALING ==========")
    print(f"Plane Scale: {plane_scale} x {plane_scale} x 1.0")
    print(f"Cube Scale: {cube_scale_xy} x {cube_scale_xy} x {cube_scale_z}")
    print(f"(Base mesh is 100x100 units, scaled to {cell_size}x{cell_size})")
    
    print("\n========== MESH INSTANCES TO SPAWN ==========")
    print(f"Floor meshes (planes): {actual_floors} instances")
    print(f"Wall meshes (cubes): {actual_walls} instances")
    print(f"Total mesh instances: {actual_floors + actual_walls}")
    
    print("\n========================================")
    
    if floor_test_passed and wall_test_passed and grid_test_passed:
        print("\n*** ALL TESTS PASSED ***")
        print("The spawning logic should generate:")
        print(f"- {actual_floors} floor plane instances")
        print(f"- {actual_walls} wall cube instances")
    else:
        print("\n*** SOME TESTS FAILED ***")
    
    return floor_test_passed and wall_test_passed and grid_test_passed

if __name__ == "__main__":
    success = test_spawning()
    exit(0 if success else 1)